feat: add LIC4 and related tests

ErikBetzholtz opened the pull request:
Closes #16

This merge will add the LIC4 function and related tests. LIC4 uses assert for illegal arguments

ErikBetzholtz committed:
feat: add LIC4 funciton and related tests

This will add the LIC4 function and related tests

ErikBetzholtz committed:
doc: add in-code documentation for LIC4

commit will add in-code documentation for LIC4 function

VincentLagerros commented on src/main/java/se/kth/EvaluateLIC.java:
```diff
@@ -29,6 +31,48 @@ public boolean LIC3(Point2D[] coordinates, double areaThreshold) {
+
+        HashSet<Integer> h = new HashSet<Integer>();
+
+        for(int i=0;i<=coordinates.length-clusterSize;i++){
```
I guess this works, but seams really inefficient when you can just do a moving window without any allocation.

ErikBetzholtz replied:
I thought so too but this was apparently a recommended solution from multiple sources

VincentLagerros replied:
What do you mean "recommended solution from multiple sources"?

ErikBetzholtz replied:
a recommended solution to finding unique elements in a list from
[here](https://stackoverflow.com/questions/13429119/get-unique-values-from-arraylist-in-java)
and
[here](https://www.baeldung.com/java-unique-values-arraylist)

VincentLagerros replied:
That is for the general use case, I would much rather use something like 
```java
        int[] quadrants = {0,0,0,0};
        int sum = 0;

        // initialize window
        for (int i = 0; i < clusterSize; i++) {
            int add = getQuadrant(coordinates[i]);
            if (quadrants[add] == 0) sum += 1;
            quadrants[add] += 1;
            if(sum>quadsThreshold) return true;
        }

        // move window
        for (int i = 0; i < coordinates.length- clusterSize; i++) {
            int remove = getQuadrant(coordinates[i]);
            int add = getQuadrant(coordinates[i+clusterSize]);
            if (quadrants[add] == 0) sum += 1;
            if (quadrants[remove] == 1) sum -= 1;

            quadrants[remove] -= 1;
            quadrants[add] += 1;
            if(sum>quadsThreshold) return true;
        }
```
with 
```java
/**
     * Get the Quadrant 0-3 mainly used for LIC4
     * @param coord The 2d position
     * @return Quadrant index, prioritizing the lower quadrant number
     */
    public static int getQuadrant(Point2D coord) {
        if(coord.getX()>=0&&coord.getY()>=0){
            return 0;
        }else if(coord.getX()<0&&coord.getY()>=0){
            return 1;
        }else if(coord.getX()<=0){
            return 2;
        }else{
            return 3;
        }
    }
```

ErikBetzholtz replied:
Sure, I'll change it

ErikBetzholtz committed:
test: add exception tests for LIC4

will add tests for invalid input to the LIC4 function to assure exceptions are thrown

ErikBetzholtz committed:
Merge branch 'main' into feature/issue#16

ErikBetzholtz committed:
minor: change order of tests in code

will change the order of tests in the EvaluateLICTest file to be in order

ErikBetzholtz committed:
Squashed commit of the following:

commit 73c6bb786f00ea700d05d0a703cf322282bb16f2
Merge: bfe99cb 5834072
Author: ErikBetzholtz <97543780+ErikBetzholtz@users.noreply.github.com>
Date:   Wed Jan 29 16:41:27 2025 +0100

    Merge pull request #40 from VincentLagerros/lic9bugfix

    Bugfix: fix LIC9 by changing

commit 5834072b7ced9aec42e97f8e5bbfdfa120e12123
Author: Vincent <vincent.lagerros@gmail.com>
Date:   Wed Jan 29 16:20:34 2025 +0100

    Bugfix: fix LIC9 by changing

    Fixes Consecutive data points interpretation

commit bfe99cb6cd6d2c9b2302dcaf700f0cb6b444aba6
Merge: 8f252dd 00dc2b2
Author: RasmussBrodin <126715286+RasmussBrodin@users.noreply.github.com>
Date:   Wed Jan 29 16:13:49 2025 +0100

    Merge pull request #38 from VincentLagerros/Fix/isssue#37

    fix: Added input test for LIC6

commit 00dc2b217d991e0a65aebbc03f2231d4e800db68
Merge: d1a9201 8f252dd
Author: RasmussBrodin <rasmuss.brodin@gmail.com>
Date:   Wed Jan 29 16:12:55 2025 +0100

    Merge branch 'main' into Fix/isssue#37

commit 8f252dd17e822f39f2a262acd37419745f48129d
Merge: 4aee3e1 ff82cc6
Author: RasmussBrodin <126715286+RasmussBrodin@users.noreply.github.com>
Date:   Wed Jan 29 16:01:52 2025 +0100

    Merge pull request #33 from VincentLagerros/Feture/issue#9

    feat: Added LIC7 and tests

commit ff82cc631fa7ebe74afdfd05d679ada2cc4fd9ff
Author: RasmussBrodin <rasmuss.brodin@gmail.com>
Date:   Wed Jan 29 15:57:18 2025 +0100

    refactor: LIC7 now uses distance()

    I did not know about distance(). The code is much cleaner now.

    Thanks!

commit 4aee3e1e9244d5805d0d14c889047c0dc152e0d7
Merge: 8238a3d 858b765
Author: ErikBetzholtz <97543780+ErikBetzholtz@users.noreply.github.com>
Date:   Wed Jan 29 15:55:21 2025 +0100

    Merge pull request #34 from VincentLagerros/feature/issue#6

    feat: Added LIC2 and test cases

commit d1a9201056fa87ce5b07c51a5b905df6b3719890
Author: RasmussBrodin <rasmuss.brodin@gmail.com>
Date:   Wed Jan 29 14:24:59 2025 +0100

    fix: Added input test for LIC6

    This closes issue #6.

    No change in code, only added test cases and it passes so should be able to merge without any problems.

commit 858b765ab45ba4b0009d10f8f8cb2a5b5bfd4c6e
Author: RasmussBrodin <rasmuss.brodin@gmail.com>
Date:   Wed Jan 29 12:23:53 2025 +0100

    Fix: Added test cases for input validation

commit 77772236a0cb77f25f9de9e788d7a52f5c9bfdc1
Author: RasmussBrodin <rasmuss.brodin@gmail.com>
Date:   Wed Jan 29 12:18:03 2025 +0100

    Fix: Added input validating test cases

commit 5d6574f41badf5ae6a5f7591d123d6b94f43a9f3
Author: RasmussBrodin <rasmuss.brodin@gmail.com>
Date:   Wed Jan 29 01:11:10 2025 +0100

    feat: Added LIC2 and test cases

    Im unsure if these test cases are sufficient, but they test the fundamental functionality

commit dd7ce34616a805e45098376106a35e975f4171c6
Author: RasmussBrodin <rasmuss.brodin@gmail.com>
Date:   Wed Jan 29 00:51:00 2025 +0100

    feat: Added LIC7 and tests

ErikBetzholtz committed:
refactor: change LIC4 to be more memory efficient

This commit will change LIC4 to use an array of size clusterSize to keep track of unique quadrants of coordinates in the moving window of coordinates. This instead of using an ArrayList of precomputed coordinates and a HashSet for checking for number of unique coordinates in window
